packages/fast-tooling-react/src/navigation/navigation.utilities.ts

```diff
@@ -1,137 +1,200 @@
 import { get } from "lodash-es";
 import { ChildOptionItem } from "../data-utilities";
-import { getDataLocationsOfChildren } from "../data-utilities/location";
-import { ItemType, TreeNavigation } from "./navigation.props";
+import { mapSchemaLocationFromDataLocation } from "../data-utilities/location";
+import { NavigationDataType, TreeNavigation } from "./navigation.props";
+import {
+    idKeyword,
+    itemsKeyword,
+    PropertyKeyword,
+    typeKeyword,
+} from "../data-utilities/types";
 
 const propsKeyword: string = "props";
 
-function getDataLocationParentLocation(
-    dataLocations: string[],
-    dataLocation: string
-): string | void {
-    const dataLocationsWithoutDataLocation: string[] = dataLocations.filter(
-        (dataLocationItem: string) => dataLocationItem !== dataLocation
-    );
-    const dataLocationParentLocations: string[] = dataLocationsWithoutDataLocation.filter(
-        (dataLocationsItem: string) =>
-            dataLocationsItem.length < dataLocation.length &&
-            dataLocation.slice(0, dataLocationsItem.length) === dataLocationsItem
-    );
+export interface NavigationFromChildrenConfig {
+    data: any;
+    schema: any;
+    dataLocation: string;
+    childOptions: ChildOptionItem[];
+}
 
-    return dataLocationParentLocations.sort(
-        (a: string, b: string) => (a.length > b.length ? -1 : 1)
-    )[0];
+/**
+ * Gets the data type from a navigation item
+ */
+function getNavigationDataType(schema: any): NavigationDataType {
+    return schema[typeKeyword]
+        ? schema[typeKeyword]
+        : schema[PropertyKeyword.properties] || schema[PropertyKeyword.reactProperties]
+            ? NavigationDataType.object
+            : schema[itemsKeyword]
+                ? NavigationDataType.array
+                : NavigationDataType.object;
 }
 
-function getNavigationFromChildLocations(
-    navigation: TreeNavigation[],
-    childrenDataLocations: string[],
+/**
+ * Gets navigation from a data location which follows
+ * lodash path syntax
+ */
+function getNavigationFromDataLocations(
     data: any,
-    childOptions: ChildOptionItem[]
+    schema: any,
+    childOptions: ChildOptionItem[],
+    dataLocation: string = "",
+    relativeDataLocation: string = "" // This data location is relative to the current schema
 ): TreeNavigation[] {
```

Comment:
You're implementation can return `undefined` - you should type this as such or adjust to always return an array.